### ISSUE TITLE
NA value for $PNR keyword crashes read.FCS

### ISSUE THREAD

**Issue by cciccole** [RGLab/flowCore#192]
**Describe the bug**
If the value NA is found for a $PNR keyword, `read.FCS()` crashes with this error:
```
Error in readFCSdata(con, offsets, txt, transformation, which.lines, scale,  : 
  $PnRNAis larger than R's numeric limit:1.79769313486232e+308
In addition: Warning message:
In readFCSdata(con, offsets, txt, transformation, which.lines, scale,  :
  NAs introduced by coercion
```

One answer to this bug may be "this isn't supposed to happen". As we know, many ways of writing FCS files are not supposed to happen but they do anyway :) I think it's reasonable to detect and work around this issue instead of exploding. Esp when `truncate_max_range` is false. Maybe set it to 0 or infer the value from an adjacent channel?

**To Reproduce**
Use this file and read it in. 
[na-range-example.fcs.zip](https://github.com/RGLab/flowCore/files/4794717/na-range-example.fcs.zip)

**Expected behavior**
The file reads in without issue.

**Comment by gfinak** [comment]
We'll aim for imputing with the real data range.

**Comment by jacobpwagner** [comment]
Yeah, I'll get on it, and apply the same logic in `load_cytoframe_from_fcs`, which actually yields an even less helpful error:
```
Error in parseFCS(normalizePath(filename), list(which.lines = which.lines,  : 
  bad lexical cast: source type value could not be interpreted as target
```

**Comment by jacobpwagner** [comment]
For common-width 32-bit and 64-bit floating point datatypes, this should be straightforward, but for integer datatypes we run in to a bit of a chicken-and-egg problem because of their variable bit-width. The range maxima are used to determine the bit mask to apply to the values read in, but if this is not known ahead of time, we will basically have to assume we can't mask any unused bits. That potentially leads to erroneous values being read, which confounds the determination of maxima for those channels. Copied from the FCS 3.1 standard, section 3.3 for easy reference:
```
For list mode storage with the binary integers data type ($MODE/L/$DATATYPE/I/), the $PnB set
of keywords specify the bit width for the storage of each parameter. The $PnR set of keywords
specify the channel number range for each parameter. For example, $P1B/16/ $P1R/1024/
specifies a 16-bit field for parameter 1 and a range for the values of parameter 1 from 0 to 1023,
which corresponds to 10 bits. Implementors should use a bit mask when reading these list mode
parameter values to insure that erroneous values are not read from the 6 unused bits. If $PnR is
not a power of 2, the next higher power of 2 should be used to create the bit mask and any
information in the bits above the next higher power of 2 should be ignored.
```
The relevant chunk of logic in `flowCore` that does this:
https://github.com/RGLab/flowCore/blob/7558d258350d314426827e419dcdc7014f98f5ec/R/IO.R#L1173-L1198

So, for invalid $PnR for integral data types, we could keep the hard error, albeit with a slightly more informative error message. We could also zero out the values in those problematic channels and throw a warning, so as to allow parsing of the other channels. Thoughts?

**Comment by gfinak** [comment]
For integral data with non-compliant PnR, I think we just fail with an error letting the user know that we cannot reliably read their non-compliant file from instrument vendor X (which we can look up) and tell them to complain to their vendor. It's on the vendor to write compliant files.

**Comment by jacobpwagner** [comment]
After https://github.com/RGLab/flowCore/commit/4dc16fa416c99d1acf3b35210ecd8c9afd62ad5f, it should parse successfully (with a warning) for missing `$PnR` on float or double datatypes. As discussed offline with @mikejiang , we will soon be switching `read.FCS` over to wrapping the C-level parser in `cytolib` currently used by `load_cytoframe_from_fcs`. For this reason I didn't spend too much time casing out the warning message as this logic will need to be rewritten there anyway.

```
> fr <- read.FCS(fcsfile)
Warning message:
In readFCSdata(con, offsets, txt, transformation, which.lines, scale,  :
  $PnR NA is invalid. If it is a numeric string, this could be because it is larger than R's numeric limit: 1.79769313486232e+308. The assigned $PnR value will be imputed from the maximum value of the data in this channel.
> fr
flowFrame object '813957a2-1b21-414b-9a8a-621a397d8527'
with 50167 cells and 8 observables:
            name          desc  range minRange maxRange
$P1        FSC-A          <NA> 262143        0   262142
$P2        FSC-W          <NA> 262144        0   262143
$P3        SSC-A          <NA> 262144        0   262143
$P4       FITC-A      CD4 FITC 262144     -111   262143
$P5         PE-A        CD8 PE 262144     -111   262143
$P6 PerCP-Cy55-A CD3 xxxxxxxxx 262144     -111   262143
$P7      Ax647-A      pErk 647 262144     -111   262143
$P8         Time          <NA> 262144        0   262143
137 keywords are stored in the 'description' slot
```
Also, I verified that the weird `xxxxxxxxx` marker for `$P6` is actually in the FCS and not some bizarre artifact added by our code...

**Comment by cciccole** [comment]
Nicely done. This is a good solution and with a great turnaround time.

Heh, the `xxxxxxxxx` was my doing since I created this example by manual edit. I deleted some of an original $PNR and because of the law of equivalent exchange within FCS text segments, I had to add text somewhere else.

In the actual example the whole data column was NA values which probably explains why the $PNR became NA. It's not clear why the entire column was NA, though.
